include/zephyr/usb/class/usb_uvc.h

```diff
@@ -0,0 +1,518 @@
+#define UVC_BMHEADERINFO_PAYLOAD_SPECIFIC_BIT	(1 << 4)
+#define UVC_BMHEADERINFO_STILL_IMAGE		(1 << 5)
+#define UVC_BMHEADERINFO_ERROR			(1 << 6)
+#define UVC_BMHEADERINFO_END_OF_HEADER		(1 << 7)
+	uint32_t dwPresentationTime; /* optional */
+	uint32_t scrSourceClockSTC; /* optional */
+	uint16_t scrSourceClockSOF; /* optional */
+	//uint8_t padding[64 - 12];
+} __packed;
+
+/** Interface Header Descriptor */
+struct uvc_interface_header_descriptor {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint16_t bcdUVC;
+	uint16_t wTotalLength;
+	uint32_t dwClockFrequency;
+	uint8_t bInCollection;
+	uint8_t baInterfaceNr;
+	/* Only a single Video Streaming interface supported for now */
+} __packed;
+
+/** Input Terminal Descriptor */
+struct uvc_input_terminal_descriptor {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint8_t bTerminalID;
+	uint16_t wTerminalType;
+	uint8_t bAssocTerminal;
+	uint8_t iTerminal;
+	uint16_t wObjectiveFocalLengthMin;
+	uint16_t wObjectiveFocalLengthMax;
+	uint16_t wOcularFocalLength;
+	uint8_t bControlSize;
+	uint8_t bmControls[3];
+} __packed;
+
+/** Output Terminal Descriptor */
+struct uvc_output_terminal_descriptor {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint8_t bTerminalID;
+	uint16_t wTerminalType;
+	uint8_t bAssocTerminal;
+	uint8_t bSourceID;
+	uint8_t iTerminal;
+} __packed;
+
+/** Camera Terminal Descriptor */
+struct uvc_camerma_terminal_descriptor {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint8_t bTerminalID;
+	uint16_t wTerminalType;
+	uint8_t bAssocTerminal;
+	uint8_t iTerminal;
+	uint16_t wObjectiveFocalLengthMin;
+	uint16_t wObjectiveFocalLengthMax;
+	uint16_t wOcularFocalLength;
+	uint8_t bControlSize;
+	uint8_t bmControls[3];
+} __packed;
+
+/** Selector Unit Descriptor */
+struct uvc_selector_unit_descriptor {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint8_t bUnitID;
+	uint8_t bNrInPins;
+	uint8_t baSourceID_iSelector[1];
+} __packed;
+
+/** Processing Unit Descriptor */
+struct uvc_processing_unit_descriptor {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint8_t bUnitID;
+	uint8_t bSourceID;
+	uint8_t wMaxMultiplier;
+	uint8_t bControlSize;
+	uint8_t bmControls[3];
+	uint8_t iProcessing;
+	uint8_t bmVideoStandards;
+} __packed;
+
+/** Encoding Unit Descriptor */
+struct uvc_encoding_unit_descriptor {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint8_t bUnitID;
+	uint8_t bSourceID;
+	uint8_t iEncoding;
+	uint8_t bControlSize;
+	uint8_t bmControls[3];
+	uint8_t bmControlsRuntime[3];
+} __packed;
+
+/** Extension Unit Descriptor */
+struct uvc_extension_unit_descriptor {
+	uint8_t bLength;
+	uint8_t bDescriptorType;
+	uint8_t bDescriptorSubtype;
+	uint8_t bUnitID;
+	uint8_t guidExtensionCode[16];
+	uint8_t bNumControls;
+	uint8_t bNrInPins;
+	uint8_t baSourceID[1];
```

Comment:
Currently there only a simple fixed chain "Input Terminal -> Output Terminal".
This is a minimal enabler without any custom controls supported, besides selecting the format in use.
This needs to be improved to support the standard as it is.

I was not sure about how to handle variable struct definitions, but this now shows me some possible way forward to fix this, thank you.

Reply:
I could use some advise from @jfischer-no and @tmon-nordic on that point:

> Descriptors are generated as `uint8_t` array initializer list because there is variable number of entities in between standard USB interface and endpoint descriptors. **Because the descriptors are automatically generated there is no real point in trying to generate corresponding C structures** (especially that some entity descriptors have variably sized fields in the middle of descriptor, e.g. Clock Selector). -- 4f1b5b11b27fe27fca6fdee8b97e86011022775d

Currently I use the structs to convert between `bFormatIndex` and `bFrameIndex`given by the host, and the actual `->wWidth`and `->wHeight` selected, but it is easy to refactor to use an `uint8_t` array and no struct.

Reply:
In UAC2 implementation I just use `uint8_t array` because the actual class does not really parse the descriptors. For the mapping the macrobatics generate simple lookup tables:
https://github.com/zephyrproject-rtos/zephyr/blob/34206f5505b582942ae4c393d7133a5ea6c849bd/subsys/usb/device_next/class/usbd_uac2.c#L746-L759
https://github.com/zephyrproject-rtos/zephyr/blob/34206f5505b582942ae4c393d7133a5ea6c849bd/subsys/usb/device_next/class/usbd_uac2.c#L791-L820

This allows the class implementation to be free of descriptor parsing and completely eliminates the variable-element-in-middle-of-structure problem.

Reply:
Thank you for the indication @tmon-nordic. I have converted UVC to use this declaration style: nothing in `include/zephyr/usb/class`.

There are slight variations to reflect the differences of UVC, and others are just due to skew of how I understand macros. Glad to hear where I could have been doing things in an unexpected way.
My goal is to make this look as familiar as possible to USB maintainers, while taking the burden of maintenance as much as I can.

Reply:
> @XenuIsWatching extenstion unit for UVC can have a variable size for the baSourceID
Extension Units are control interfaces, and control interfaces are not yet supported, but control interface *descriptors* are now supported, including Extension Unit with several `baSourceID`s.

This now needs support for control commands, and then it would be possible to use custom controls.
I thought of mapping them to these:
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/video-controls.h#L40

For instance, if the goal is to combine two image sensors into a custom pixel format and expose custom controls over USB to some custom block doing the combination:

```
                /* Control chain from the 2 image sensors to the output */
                uvc0cam0: cam0 {
                        compatible = "zephyr,uvc-control-camera";
                        control-target = <&imx477>; /* RGB */
                        control-ids = <UVC_CONTROL_EXTENSION_>;
                };
                uvc0cam1: cam1 {
                        compatible = "zephyr,uvc-control-camera";
                        control-target = <&IMX556>; /* ToF */
                };
                uvc0ext: ext {
                        compatible = "zephyr,uvc-control-extension";
                        control-ids = <UVC_CONTROL_EXTENSION_0,
                                       UVC_CONTROL_EXTENSION_1,
                                       UVC_CONTROL_EXTENSION_2>;
                        control-target = <&custom_device>;
                        source-entity = <&uvc0cam0, &uvc0cam1>;
                };
                uvc0out: out {
                        compatible = "zephyr,uvc-control-output";
                        source-entity = <&uvc0ext>;
                };

                /* Custom format definition */
                format_uncompressed {
                        compatible = "zephyr,uvc-format-uncompressed";
                        fourcc = "XTRA";
                        guid = [OUR_CUSTOM_UVC_FORMAT_GUID];
                        bits-per-pixel = <24>;

                        FHD {
                                size = <1920 1090>;
                                max-fps = <60>;
                        };
                };
```

This currently only works at the descriptor level, next I need to parse the control commands so they can affect the video device referred by the matching `control-target`.